www/js/controller.js

```diff
@@ -276,6 +277,78 @@ define([
 			}));
 		},
 
+		checkDone: function () {
+			/*
+				When the author clicks on the main menu "Done" button,
+				the system checks that one variable is Root, 
+				and make sure every variable is part of at least one equation.
+			*/
+			console.log("Done Action called");
+			var returnObj = {};
+			returnObj.errorNotes = [];
+			var hasRootNode = false;
+			var _variables = [];
+			var _equations = [];
+			
+			if( this._model ){
+				array.forEach(this._model.active.getNodes(), function (node) {
+					if(node.root){
+						hasRootNode = true;
+					}
+					console.log(node);
+
+					// get variables and equations
+					if(node.variable != "" && node.genus && node.genus == "required" && node.type == "quantity"){
+						_variables.push(node.ID);
+					}else if(node.type == "equation"){
+						_equations.push(node.equation);
+					}
+
+				});
+
+				// check if each variable is present in atleast one equation
+				var _usedVariables = _variables.map(function(_variable){
+					var tmp = array.some(_equations, function(_equation) {
+						
+						if( _equation && _equation.search(_variable) > -1){
+							return true;
+						};
+					});
+					var obj = {};
+					obj[_variable] = tmp;
+					return obj;
+				});
+
+				// for each variable if it is not present in any equation, add it to array 
+				var _requiredVariables = [];
+				array.forEach(_usedVariables, function(item){
+					
+					if(!Object.values(item)[0]){	
+						_requiredVariables.push(Object.keys(item)[0]);
```

Comment:
A variable is required if the author set its optionality to required.  Otherwise it is not.  We don't seem to be checking its optionality here though, so unless I'm missing something I think there's a bug here where optional nodes will be counted as required even though they're not.